[search] Import Tiger.

vng opened the pull request:
Closes https://github.com/organicmaps/organicmaps/issues/2532

3768 lines are OSM test data.

Has debug addr:interpolation drawing:

<img width="1309" alt="image" src="https://github.com/organicmaps/organicmaps/assets/175612/01922a6b-4bbe-4b37-bbde-7242a46158b1">

biodranik commented on coding/read_write_utils.hpp:
```diff
@@ -5,24 +5,38 @@
 {
-  template <class TSink>
-  void Write(TSink & sink, uint32_t i)
+  template <class T, class TSink>
+  std::enable_if_t<std::is_integral<T>::value && std::is_unsigned<T>::value, void>
+  Write(TSink & sink, T i)
```
Will this work?

```
   template<class T, class TSink>
   requires std::is_integral_v<T> && std::is_unsigned_v<T>
   Write(TSink & sink, T i)
```

biodranik commented on coding/coding_tests/reader_writer_ops_test.cpp:
```diff
@@ -119,3 +120,35 @@ UNIT_TEST(ReadWrite_POD)
+{
+template <class T> void TestIntegral()
+{
+  std::vector<T> ethalon{ static_cast<T>(-1), 0, 1,
```
etalon

vng replied:
"ethalon" here.
Anyway, there are no Eng words like "etalon" or "ethalon", but Ethalon is used in string matching.

biodranik commented on coding/read_write_utils.hpp:
```diff
@@ -5,24 +5,38 @@
+  std::enable_if_t<std::is_integral<T>::value && std::is_unsigned<T>::value, void>
+  Write(TSink & sink, T i)
   {
     WriteVarUint(sink, i);
```
Does WriteVarUint(sink, int8_t i) make any sense?

biodranik replied:
Should this overload be disabled/overridden?

biodranik commented on indexer/indexer_tests/search_string_utils_test.cpp:
```diff
@@ -288,4 +288,24 @@ UNIT_TEST(NormalizeAndSimplifyString_Apostrophe)
+  };
+
+  {
+    std::string_view const ethalon = "680northwest";
```
etalon

biodranik commented on indexer/indexer_tests/search_string_utils_test.cpp:
```diff
@@ -288,4 +288,24 @@ UNIT_TEST(NormalizeAndSimplifyString_Apostrophe)
+  {
+    std::string_view const ethalon = "680northwest";
+    Check("N 680 W", ethalon);
+    Check("North 680 West", ethalon);
```
What about `N 680 West`?

biodranik commented on indexer/indexer_tests/search_string_utils_test.cpp:
```diff
@@ -288,4 +288,24 @@ UNIT_TEST(NormalizeAndSimplifyString_Apostrophe)
+
+  Check("North 20th Rd", "20thnorth");
+  Check("Lane st", "lane st");         /// @todo Probably, doesn't matter here?
+  Check("West North", "westnorth");    /// @todo Should order?
```
Usually names start with north or south.

zyphlar replied:
We have a perplexing variety. I just edited West North Street, which is not the same thing as Northwest ___ Street. Many cities will have suffixes instead, like 1st Street [North/South] or Main Street NE. It's unfortunately almost impossible to generalize.

biodranik commented on indexer/search_string_utils.cpp:
```diff
@@ -402,6 +402,36 @@ class StreetsSynonymsHolder
+    for (auto const * s : {"n", "north", "s", "south", "w", "west", "e", "east",
+                           "ne", "northeast", "nw", "northwest", "se", "southeast", "sw", "southwest" })
+    {
+      m_strings.emplace_back(NormalizeAndSimplifyString(s));
```
1. Is normalization and simplification really needed here?
2. Does it make sense to add a UniString(char32_t const * u32str) constructor and initialize like this:

```
m_strings {U"n", U"north", ...}
```

biodranik commented on indexer/search_string_utils.cpp:
```diff
@@ -402,6 +402,36 @@ class StreetsSynonymsHolder
+    }
+  }
+
+  template <class FnT> bool ApplyIf(UniString const & s, FnT && fn) const
```
It would be clearer to put template on a separate line.

biodranik commented on generator/CMakeLists.txt:
address_parseR_tool in the commit name?

biodranik commented on generator/address_parser/parser_tool.cpp:
```diff
@@ -0,0 +1,31 @@
+
+MAIN_WITH_ERROR_HANDLING([](int argc, char ** argv)
+{
+  std::string usage("Tiger addresses processor. Sample usage:\n");
```
const

biodranik commented on generator/address_parser/processor.cpp:
```diff
@@ -0,0 +1,93 @@
+  {
+    ++total;
+
+    m_workers.SubmitWork([this, &incomplete, copy = std::move(line)]() mutable
```
How big was the speedup compared to single-thread processing without copying strings and mutex?

biodranik commented on generator/address_parser/tiger_parser.cpp:
```diff
@@ -0,0 +1,132 @@
+    CHECK(i != std::string_view::npos, (s));
+
+    CHECK(to_double(s.substr(0, i), last.m_lon), (s));
+    CHECK(to_double(s.substr(i+1), last.m_lat), (s));
```
nit: `i + 1`

biodranik commented on generator/address_parser/tiger_parser.cpp:
```diff
@@ -0,0 +1,132 @@
+  e.m_to = *it; ++it;
+  e.m_interpol = ParseInterpolation(*it); ++it;
+  e.m_street = *it; ++it;
+  ++it; ++it;
```
What is skipped here? A comment?

biodranik commented on generator/address_enricher.hpp:
```diff
@@ -0,0 +1,85 @@
+
+    template <class TSink> void Save(TSink & sink) const
+    {
+      rw::Write(sink, m_from);
+      rw::Write(sink, m_to);
+      rw::Write(sink, m_street);
+      rw::Write(sink, m_postcode);
```
Any chance that this sequence will change when open addresses are supported?

biodranik replied:
What is the format used for Open Addresses? Should it be taken into an account here to avoid `if (v1) {} else { // process openaddresses }` ?

biodranik reviewed (commented):
It would be great to merge several first commits separately.

vng commented on base/random.hpp:
```diff
@@ -1,10 +1,22 @@
+{
+  static_assert(std::is_integral<T>::value);
+
+  std::random_device m_rd;
```
I added just a syntax sugar without changing the logic.

biodranik replied:
I understand. If the generator is created too often, please add a TODO to fix it later by reusing the generator.

biodranik commented on indexer/search_string_utils.cpp:
```diff
@@ -432,6 +428,44 @@ class StreetsDirectionsHolder
+  StreetsDirectionsHolder()
+  {
+    // ("short name", "full name")
+    for (auto const * s : {"n", "north", "s", "south", "w", "west", "e", "east",
```
These shortcuts are very specific to countries and languages. E.g. they should not be applied in German or French-speaking countries, for example. Is there any plan to take it into account? Any specific issue?

zyphlar replied:
FYI @vng here are some other similar lists that may help complete this for the US

https://wiki.openstreetmap.org/wiki/Seattle_Import#SeattleAddress_Transformation

https://git.zyphon.com/will/the-villages-import/src/branch/main/qgis-functions.py#L32

biodranik commented on indexer/search_string_utils.cpp:
```diff
@@ -466,21 +500,46 @@ UniString GetStreetNameAsKey(std::string_view name, bool ignoreStreetSynonyms)
 
+strings::UniString GetNormalizedStreetName(std::string_view name)
+{
+  static StreetsDirectionsHolder s_directions;
```
Do statics here produce unnecessary multithreading overhead? Would it be better to make them a member of a class instance?

vng commented:
The first data snapshot with US_* sizes diff https://pastebin.com/cep9hkWu

vng commented:
A quick look to NY "E 14th St". Despite of some heuristics to filter addr:interpolarion Tiger data if we _already have_ node addresses from OSM, got this picture (blue dash lines shows imported addr:interpolation):
![image](https://github.com/organicmaps/organicmaps/assets/175612/35faf02b-a601-4555-a9a0-debfb81c5de5)

vng commented:
SF, "Birch St". But in general OSM SF addresses are poor, so addr:interpolation looks good there:
![image](https://github.com/organicmaps/organicmaps/assets/175612/f5a81247-e078-4e7a-bf9e-7657a9dcd8f5)

vng commented:
The address picture is better in _the middle of nowhere_:

![image](https://github.com/organicmaps/organicmaps/assets/175612/8a4773e6-3a6c-4379-a34f-f9c97d28feeb)

biodranik commented:
1. That's a huge increase. Can the size change be minimized?
<img width="613" alt="выява" src="https://github.com/organicmaps/organicmaps/assets/170263/66979915-db1a-4f1b-8b03-14ac374743d2">


2. 
> Despite some heuristics to filter addr:interpolarion Tiger data if we already have node addresses from OSM, got this picture

2.1 Will the search show two results in this case?
2.2 How to filter out this duplicating interpolation data?

vng commented:
Latest stats:
```
Total addresses: { m_noStreet = 4365877; m_existInterpol = 29668; m_existSingle = 81686; m_enoughAddrs = 2785897; m_addedSingle = 851968; m_addedBegEnd = 51175297; m_addedInterpol = 25860658 }
```

- 4.3M addresses - correspondent street not found in OSM
- Added new 25.8M interpolation addresses with 51M beg/end nodes
- Added 0.85M single node range addresses (like 110-120)

biodranik commented:
@vng what's needed to finish it? Can the generator data be ready on Sunday?

vng commented:
Updated sizes diff with bigger _exist_ threshold https://pastebin.com/eP8RSV79

vng commented:
Tiger has some strange (duplicate) entries, like:
```
4701;4799;all;E Mc Kinley Ave;Fresno;CA;93727;LINESTRING(-119.737784 36.765257,-119.737654 36.765254,-119.736474 36.765244)
4701;4799;odd;E Mc Kinley Ave;Fresno;CA;93703;LINESTRING(-119.740705 36.765261,-119.738210 36.765256)
```

biodranik commented:
> Tiger has some strange (duplicate) entries, like:

It is done to match two different zip codes with the same addresses. Does our search support more than one zip code for the same address?

vng commented:
They are in different blocks

vng commented:
Latest stats with threshold = 50m:
```
Total addresses: { m_noStreet = 4299981; m_existInterpol = 30587; m_existSingle = 107684; m_enoughAddrs = 4215096; m_addedSingle = 833089; m_addedBegEnd = 48739445; m_addedInterpol = 24489317 }
```

biodranik commented:
@vng is it hard to make a table? What do these values mean?

zyphlar commented:
@vng FYI the other difference with those lines is even vs odd vs all: addresses 4701, 4703, 4705, etc will be the "odd" side of the road, whereas the other side will be the same range but even.

These GPS coordinates are in a very small area, barely a whole city block, so I agree it's probably leftover junk. TIGER data is relatively low quality.

zyphlar commented:
Can you make a list (geojson?) of TIGER address interpolation lines whose streets don't match, maybe we can do some QA?

Ideally OM would support address interpolation lines in OSM directly, and we could prioritize that OSM data (fixed vs TIGER) over TIGER itself, for easy bugfixing. (No sense in waiting for TIGER to fix itself.)

Osyotr commented on coding/coding_tests/reader_writer_ops_test.cpp:
```diff
@@ -39,11 +40,11 @@ namespace
     FileWriter writer(fName);
 
-    srand(666);
+    base::UniformRandom<int8_t> rand;
```
@vng fyi https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution
> The effect is undefined if this is not one of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long.

biodranik replied:
@Osyotr was it highlighted by VS?

Osyotr replied:
There's a `static_assert` for that in MSVC STL.
https://github.com/microsoft/STL/blob/a357ff1750d3f6dffd54b10d537e93e0accfcc92/stl/inc/random#L2340

brdbkr commented:
https://github.com/organicmaps/organicmaps/issues/9252

This was a bad change - I certainly don't expect an alleged OSM app to also include non-OSM data.

TIGER is an imperfect source at best. See the [OSM Wiki TIGER page](https://wiki.openstreetmap.org/wiki/TIGER):

> However, while the import helped kickstart OSM data in the US, it also brought a plethora of problems with data inaccuracy, some of which persist to this day. Lots of cleanup has been done from the initial import.

zyphlar commented:
@brdbkr I promise it's not a bad change: I've been trying to increase adoption of any open source gps app for almost ten years and with this fix I was finally able to search for an address I hadn't personally imported into osm and get results. OSM.org itself uses TIGER data to search for addresses not present in the database, via Nominatim. The sad fact is that even with import efforts like mine, address coverage is extremely low nationally and globally. We need other data sources in order to make the app a usable replacement for closed-source apps.

The way it's implemented can of course be improved! But OSM itself will tell you, it's not OSM's job to provide all the data necessary for good GPS apps (like, say, topographical contours, time zones, country to language code associations, or additional data like Wikidata and Wikipedia.) Apps are expected to correlate multiple data sources, not try and shove information into OSM just because they want an "everything map."

One way of fixing this for you could be to hide the numbers on the map itself. Then TIGER would only appear in search results. OM could also mark search results with the data source to reduce confusion when data is inaccurate.

brdbkr commented:
> One way of fixing this for you could be to hide the numbers on the map itself. Then TIGER would only appear in search results.

That would help a lot I think!
